R/teal_lockfile.R

```diff
@@ -23,92 +44,140 @@
+  checkmate::assert_string(lockfile_path)
+  checkmate::assert_list(opts)
+  checkmate::assert_class(sysenv, "list")
+  checkmate::assert_character(libpaths, min.len = 1)
+  checkmate::assert_directory(wd)
 
-    renv_logs <- utils::capture.output(
-      renv::snapshot(
-        lockfile = lockfile_path,
-        prompt = FALSE,
-        force = TRUE
-        # type = is taken from renv::settings$snapshot.type()
-      )
+  # mirai starts in vanilla session in the R.home directory. We need to pass all session related info
+  options(opts)
+  lapply(names(sysenv), function(sysvar) do.call(Sys.setenv, sysenv[sysvar]))
+  .libPaths(libpaths)
+  setwd(wd)
+
+  out <- utils::capture.output(
+    renv <- renv::snapshot(
+      lockfile = lockfile_path,
+      prompt = FALSE,
+      force = TRUE
+      # type is taken from renv::settings$snapshot.type()
     )
-    if (any(grepl("Lockfile written", renv_logs))) {
-      logger::log_trace("lockfile created successfully.")
+  )
+
+  list(
+    out = out,
+    lockfile_path = lockfile_path,
+    length = length(renv$Packages)
+  )
+}
+
+#' @rdname teal_lockfile
+#' @keywords internal
+teal_lockfile_tracker <- function(process) {
+  checkmate::assert_class(process, "ExtendedTask", null.ok = TRUE)
+  if (inherits(process, "ExtendedTask")) {
+    tracker <- observeEvent(process$status(), {
+      if (process$status() != "running") {
+        teal_lockfile_handler(process)
+      }
+    })
+  } else {
+    # if it is not mirai then lockfile was provided through options (skip the process - file is ready)
+    shinyjs::show("teal-lockFile")
+  }
+}
+
+#' @rdname teal_lockfile
+#' @keywords internal
+teal_lockfile_handler <- function(process) {
+  # todo: make sure that we catch all possible status outputs
+  #     - normally getwd should be set to the app directory (snapshot based on the app files)
+  #     - what happens if app directory is set to a different location?
+  #     - what if setwd is set to a directory which doesn't contain anything
+  #     - check out possible ERROR and WARNING
+  renv_logs <- process$result()
+  if (any(grepl("Lockfile written to", renv_logs$out))) {
+    with <- if (any(grepl("WARNING:", renv_logs$out))) {
+      " with warning(s)"
+    } else if (any(grepl("ERROR:", renv_logs$out))) {
+      " with error(s)"
     } else {
-      logger::log_trace("lockfile created with issues.")
+      ""
     }
 
-    lockfile_path
-  })
-  if (close) {
-    # If the initial backend was only sequential, bring it back.
-    promises::then(promise, onFulfilled = function() {
-      future::plan(future::sequential)
-    })
+    logger::log_trace("Lockfile {renv_logs$lockfile_path} containing { renv_logs$length } packages created{ with }.")
+    shiny::showNotification("Lockfile is available to download.")
+    shinyjs::show("teal-lockFile")
+  } else {
+    warning("Lockfile creation failed.")
+    shiny::showNotification("Lockfile creation failed.", type = "warning")
```

Comment:
I would show it anyway with a NOTE, that you download the file on your own risk, since the file was created but `renv::restore` might not be able to restore project based on this lockfile.

```suggestion
    shiny::showNotification("Lockfile creation failed.", type = "warning")
    shinyjs::show("teal-lockFile")
```